packages/monaco/src/browser/monaco-quick-input-service.ts

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+import { MonacoQuickOpenService } from "./monaco-quick-open-service";
+import { injectable, inject } from "inversify";
+import { ILogger } from "@theia/core/lib/common/logger";
+import { QuickOpenItem, QuickOpenItemOptions, QuickOpenMode } from "@theia/core/lib/browser/quick-open";
+
+export interface QuickInputOptions {
+
+    /**
+     * The prefill value.
+     */
+    value?: string;
+
+    /**
+     * The text to display under the input box.
+     */
+    prompt?: string;
+
+    /**
+     * The place holder in the input box to guide the user what to type.
+     */
+    placeHolder?: string;
+
+    /**
+     * Set to `true` to show a password prompt that will not show the typed value.
+     */
+    password?: boolean;
+
+    /**
+     * Set to `true` to keep the input box open when focus moves to another part of the editor or to another window.
+     */
+    ignoreFocusOut?: boolean;
+
+    /**
+     * An optional function that will be called to validate input and to give a hint
+     * to the user.
+     *
+     * @param value The current value of the input box.
+     * @return Return `undefined`, or the empty string when 'value' is valid.
+     */
+    validateInput?(value: string): string | undefined | PromiseLike<string | undefined>;
+}
+
+@injectable()
+export class MonacoQuickInputService extends MonacoQuickOpenService {
```

Comment:
OK, but I also need way to control [onFocusLost](https://github.com/theia-ide/theia/blob/master/packages/monaco/src/browser/monaco-quick-open-service.ts#L72) exit.
Should I expose `setPassword` an `ignoreFocusOut` option to the general Theia APi or I can just modify `MonacoQuickOpenService` and use it directly?

Reply:
And I have question about [this function](https://github.com/theia-ide/theia/blob/master/packages/monaco/src/browser/monaco-quick-open-service.ts#L60) do this function do not return focus to previously focused html element by design or it's just bug?

by design
--




by design
--

Reply:
> OK, but I also need way to control onFocusLost exit.
Should I expose setPassword an ignoreFocusOut option to the general Theia APi or I can just modify MonacoQuickOpenService and use it directly?

It looks like for this PR modifying `MonacoQuickOpenService` should be enough. We can reconsider core API with follow-up PRs, maybe even `QuickInputService` implemented with `QuickOpenService` could go eventually to core.

I would try to avoid exposing `monaco.quickOpen.QuickOpenWidget` itself, since it is internals of VS Code and a subject to change. It would be nice if we need to change only `MonacoQuickOpenService` if it changes without worrying for other extensions.

> And I have question about this function do this function do not return focus to previously focused html element by design or it's just bug?
 
sorry, i could not get what the issue is. Could you elaborate how i can reproduce from the user perspective?

We reset the focus back if a quick widget is canceled. Do you mean that we should also reset the focus back is something was chosen? In some cases a chosen item can open a new widget in that case resetting the focus back will be bogus. For example a command to open an editor for a diff editor.

Reply:
>It looks like for this PR modifying MonacoQuickOpenService should be enough. We can reconsider core API with follow-up PRs, maybe even QuickInputService implemented with QuickOpenService could go eventually to core.

>I would try to avoid exposing monaco.quickOpen.QuickOpenWidget itself, since it is internals of VS Code and a subject to change. It would be nice if we need to change only MonacoQuickOpenService if it changes without worrying for other extensions.

OK, I will do in that way.
>We reset the focus back if a quick widget is canceled. Do you mean that we should also reset the focus back if something was chosen? In some cases a chosen item can open a new widget in that case resetting the focus back will be bogus. For example a command to open an editor for a diff editor.